### ISSUE TITLE
[DOXIA-554] Parsing Markdown documents can hang site generation: switch parser from Pegdown to Flexmark

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-doxia#527]
**[Michael Benz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=focbenz)** opened **[DOXIA-554](https://issues.apache.org/jira/browse/DOXIA-554?redirect=false)** and commented

The parsing time for Markdown documents can take very long and hang site generation when e.g. long tables are being generated.

The author of pegdown has marked the pegdown project deprecated since 2016-12-14 [pegdown.org](https://github.com/sirthias/pegdown/) and advises to switch to [flexmark-java](https://github.com/vsch/flexmark-java).

> The project is essentially unmaintained with tickets piling up and crucial bugs not being fixed.
> pegdown's parsing performance isn't great. In some cases of pathological input runtime can even become exponential, which means that the parser either appears to "hang" completely or abort processing after a time-out.

Since the parsing timeout was increased in DOXIA-498 it is now possible to "hang" the site creation with a longer table like the one in this example.
In case this sample is rendered using version 3.3 of the maven site

---

**Affects:** 1.7

**Attachments:**
- [maven-pom-sample-pegdown-performance.zip](https://issues.apache.org/jira/secure/attachment/12850848/maven-pom-sample-pegdown-performance.zip) (_4.68 kB_)

**Issue Links:**
- [DOXIA-571](https://issues.apache.org/jira/browse/DOXIA-571) Markdown does not support syntax highlighting any more
 (_**"causes"**_)
- [DOXIA-553](https://issues.apache.org/jira/browse/DOXIA-553) Replace pegdown due to deprecation and end of life notice

- [DOXIA-523](https://issues.apache.org/jira/browse/DOXIA-523) Ampersands after hash in markdown link syntax are now throwing parse errors.

- [DOXIA-568](https://issues.apache.org/jira/browse/DOXIA-568) Doxia Markdown extension > Support for some MarkDown extensions has been removed since 1.7

- [MSITE-791](https://issues.apache.org/jira/browse/MSITE-791) upgrade Doxia to 1.8 to benefit from new Markdown parser (require Java 7)
 (_**"is required by"**_)
- [DOXIA-567](https://issues.apache.org/jira/browse/DOXIA-567) update dependency schema: pegdown not used any more but flexmark-java
 (_**"is depended upon by"**_)

1 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

The default, inherited Java 6 `javac` compiler version can be overridden locally in doxia-module-markdown.

https://maven.apache.org/plugins/maven-compiler-plugin/

The following addition would override Java 6 compiler with the Java 7 compiler.

_doxia-modules/doxia-module-markdown/pom.xml_

```xml
<project>
  [...]
  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <configuration>
          <source>1.7</source>
          <target>1.7</target>
        </configuration>
      </plugin>
    </plugins>
  </build>
  [...]
</project>
```

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

I was referring to maven-doxia version. The pom.xml for maven-doxia has:

```
<groupId>org.apache.maven.doxia</groupId>
<artifactId>doxia</artifactId>
<version>1.7</version>
<packaging>pom</packaging>
```

Changing the JDK version for the module is not a problem. I don&rsquo;t know how to direct the PR towards the right maven-doxia version since GitHub mirror only has the trunk version 1.7

The git://git.apache.org/maven-doxia.git is exactly the same as the GitHub repo.

BTW, the change of the Markdown module to flexmark-java, with Java language level 1.7, is complete and passing tests. The only hitch is what version to assign to it and figuring out where to commit the changes an generate a PR.

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

You can get the changes from https://github.com/vsch/maven-doxia where I modified the doxia-module-markdown for flexmark-java parser.

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

I spoke prematurely. Still getting the enforcer failing on byte code version:

```
[INFO] ------------------------------------------------------------------------
[INFO] Building Doxia :: Markdown Module 1.8-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-enforcer-plugin:1.4.1:enforce (enforce-bytecode-version) @ doxia-module-markdown ---
[INFO] Restricted to JDK 1.6 yet org.nibor.autolink:autolink:jar:0.6.0:compile contains org/nibor/autolink/Autolink.class targeted to JDK 1.7
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.EnforceBytecodeVersion failed with message:
Found Banned Dependency: org.nibor.autolink:autolink:jar:0.6.0
Use 'mvn dependency:tree' to locate the source of the banned dependencies.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.101 s
[INFO] Finished at: 2017-03-08T17:52:41-05:00
[INFO] Final Memory: 23M/464M
[INFO] ------------------------------------------------------------------------
```

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch) The following can be added to "doxia-modules/doxia-module-markdown/pom.xml" to set the enforcer enforceBytecodeVersion to 1.7 for the doxia-module-markdown.

```xml
<plugins>
  <plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-enforcer-plugin</artifactId>
    <executions>
      <execution>
        <id>enforce-bytecode-version</id>
        <goals>
          <goal>enforce</goal>
        </goals>
        <configuration>
          <rules>
            <enforceBytecodeVersion>
              <maxJdkVersion>1.7</maxJdkVersion>
            </enforceBytecodeVersion>
          </rules>
          <fail>true</fail> 
        </configuration>
      </execution>
    </executions>
  </plugin>        
  <plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-compiler-plugin</artifactId>
    <!-- ... ->
```

I checked a build with the addition of the above code to https://github.com/vsch/maven-doxia. After the enforcer enforceBytecodeVersion is updated to maxJdkVersion 1.7, the next apparent hurdle is licensing.

The automatic license checker `org.apache.rat:apache-rat-plugin:0.11:check (rat-check)` found the licenses to be "Unapproved: 2 unknown: 2 generated: 0 approved: 8".

Two files in question appear to be ones that Vladimir, as the original code author and copyright holder, could adjust. `MarkdownDoxiaNodeRenderer.java` and `MarkdownDoxiaExtension.java` have:

```java
/* Copyright (c) 2015-2016 Vladimir Schneider <...>, all rights reserved.
 *
 * This code is private property of the copyright holder and cannot be used without
 * having obtained a license or prior written permission of the of the copyright holder.
```

From the report it appears the 2 unknown licenses are also the unapproved licences. Building the module with the above enforcer change will provide the `rat.txt` licensing report.

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

Thanks Marc. The plugin config change works. The license replaced with ASF 2.

I did not setup a copyright config in IntelliJ so when I copied the code over from another project it left the copyright as is.

I now have the build and test running but need to figure out the rendered HTML differences because the tests do not match what pegdown generated. Looking into it now.

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)  The `doxia-modules/doxia-module-markdown/src/site/apt/index.apt` should inform the users of considerations in the choice between `doxia-module-markdown` version 1.7 and 1.8 because of the different required Java version.

My first thoughts on additional text would be something like the following.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), [Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) Comments|input|guidance|feedback are welcome. :-)

---

**NOTE: Unlike other `Doxia` 1.8 modules, `doxia-module-markdown` version 1.8 requires Java 7.  Other `Doxia` 1.8 modules require Java 6.**

The `doxia-module-markdown` version can be specified separately from other Doxia modules.

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-site-plugin</artifactId>
    <version>${maven-site-plugin.version}</version>
    <dependencies>
        <dependency>
            <groupId>org.apache.maven.doxia</groupId>
            <artifactId>doxia-module-markdown</artifactId>
            <version>${doxia-module-markdown.version}</version>
        </dependency>
        <!-- other doxia module(s) -->
    </dependencies>
</plugin>
```

If your project requires Java 6, then you may consider the use of `doxia-module-markdown` version 1.7.  In this situation, you should be aware that `pegdown` has known technical issues and that ["pegdown has reached its end of life"](http://pegdown.org).

The `flexmark-java` markdown processor in `doxia-module-markdown` version 1.8 provides the following improvements.

* Modern parser architecture (based on commonmark-java), designed from the ground up as a pegdown replacement and supporting all its features and extensions
* 30x better average parsing performance without pathological input cases
* Configuration options for a multitude of markdown dialects (CommonMark, pegdown, MultiMarkdown, kramdown and Markdown.pl)

... references ...

* `[flexmark-java](http://github.com/vsch/flexmark-java)`, the library used by Doxia `doxia-module-markdown` version 1.8 in `[Pegdown](http://pegdown.org)` compatibility mode.

* `[Pegdown](http://pegdown.org)`, a deprecated markdown processing library used in the previous `doxia-module-markdown` version 1.7.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

I like it: I improved configuration and added the notice in https://github.com/vsch/maven-doxia/pull/1

But I get 2 failures when building:

```
Tests in error: 
 MarkdownParserTest>AbstractParserTest.testDocument:116 » Parse Error parsing t...
 MarkdownParserTest>AbstractParserTest.testParser:83 » Parse Error parsing the ...
```

I don't know if by default next maven-site-plugin will include 1.7 or 1.8 Markdown Doxia module: but it's great to have the choice, and just have to decide what's the default config and which users will have to dig into more complex configuration

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)  I will share my thoughts on default for your consideration...

As a long time maven user, I have two broad groups of maven projects: "legacy" and "new/not-legacy".

The "legacy" projects have expressly stated versions in an appropriate parent POM.  Version updates only occur with specific review of the version change implications. ("legacy" inheritently has deliberate, cautious version changes)

A "new" projects, as would be created from an IDE template or otherwise, are expected _to *not* default to something known to be "broken"_.  ("new" likes an automatic default to work with minimal headaches and legacy barriers.)

So, in my use case,

(1) the preferred "default" scenario (which implies "new project") would  _*not* default to a "technical issues" and "un-supported EOL" component_.  So, `flexmark-java` would be preferred to `pegdown` as the more current default.

_AND_

(2) clear documentation, easy-to-find info needed to management version selection for conscious legacy versioning. (e.g. main plugin and/or module web pages. maybe adviory statement(s) in the build output.)

Basically, in my case, "legacy" has expressly mananged versions and "new" is expected to default conveniently to a non-broken, non-limiting start.

That said, I realize other users (and this Maven Project) may have a different approaches, priorities and preferences.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

yes, I like it: the doc in index.apt is quite ready for that
now, there are 2 unit tests to fix before I can merge

**Comment by jira-importer** [comment]
**[Michael Benz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=focbenz)** commented

I like the improvements and would second Marc Campbell on the overall legacy user scenario. Would also switich to markdown doxia module 1.8 and run a few tests against our documentation.

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)  I use Netbeans for maven projects. And IntelliJ/AndroidStudio for Android/gradle projects.

For maven, I've found maven in IntelliJ to have "add-on" feel & oddities. Similar to DZone post ["My Experiences with Maven in IntelliJ IDEA and NetBeans IDE "](https://dzone.com/articles/my-experiences-maven-intellij)

For me today, https://github.com/vsch/maven-doxia in IntelliJ gives a list of project warnings which would need to be tracked down.

When I open https://github.com/vsch/maven-doxia in Netbeans, `doxia` builds and the "Debug Test File" just works. No headaches.

**Comment by jira-importer** [comment]
**[Marc Campbell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=moonrocket)** commented

Try using `mvn clean install` instead of `mvn clean package` for the multi-module build.  `install` will deploy each module to the local .m2/repository to be available for the subsequent module.  Netbeans "Clean and Build" executes `mvn clean install`

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

Thanks Marc. I got NetBeans installed and it builds and debugs out of the box. I can use it. The PITA is that there is no Markdown plugin for NetBeans 8.2 and even if there was it is underpowered. Having it running on NetBeans, I got it working on IntelliJ, so I can use my Markdown plugin to see the HTML that will be generated by the test.

I am now doing comparison between doxia 1.7 with pegdown and 1.8 with flexmark to see why there are differences because I cannot see where the differences would be from generated HTML by old pegdown and flexmark-java.

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

Modified tests and flexmark-java options to not output extra formatting EOLs, closer to pegdown's unformatted output. Now all passing.

Next hurdle:

```
[INFO] --- clirr-maven-plugin:2.7:check (default) @ doxia-module-markdown ---
[INFO] Comparing to version: 1.7
[ERROR] Unable to locate enclosing class org.apache.maven.doxia.module.markdown.MarkdownParser for nested class org.apache.maven.doxia.module.markdown.MarkdownParser$PegDownHtmlParser
[ERROR] 8001: org.apache.maven.doxia.module.markdown.MarkdownToDoxiaHtmlSerializer: Class org.apache.maven.doxia.module.markdown.MarkdownToDoxiaHtmlSerializer removed
```

The MarkdownToDoxiaHtmlSerializer should have never been public in the original markdown module. It is purely an implementation detail. I also left the PegDownHtmlParser class name as is to avoid more errors, although it should be named MarkdownHtmlParser to make it independent of the parsing library.

Is there something I am missing in the configuration to allow an API breaking change to be allowed for the module?

The alternative of leaving the dead class and dependency on pegdown does not seem like the right solution.

My fork was updated with the changes to date.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

for issues reported by clirr that you want to ignore, just add a new line in Doxia parent pom: there are already a good number of such configurations

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

Thank you Hervé. All fixed. PR generated.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

PR merged, with little updates: now, I'll do MSITE-791 to integrate the result in m-site-p 3.7

thank you for your great work

**Comment by jira-importer** [comment]
**[Vladimir Schneider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vsch)** commented

Hervé, My pleasure and thank you for guiding this newbie through doxia intricacies.

Let me know if any issues come up that need addressing.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

Jira is the issues repository: you'll need to find an issue that you find interest in

If you are really motivated, MPDF-48 can be hard but interesting: make the mavent-pdf-plugin run reports with Maven 3. This will require a lot of cleanup, digging into details of reports execution to pick the right part of maven-site-plugin

**Comment by jira-importer** [comment]
**[Michael Benz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=focbenz)** commented

No - the next version 1.8 has not been released. You would need to checkout the source and install this into your local repository.
Just checked https://mvnrepository.com/artifact/org.apache.maven.doxia/doxia-module-markdown to make sure.
